### PR TITLE
Added expose loaded blackbox configuration on /config

### PR DESCRIPTION
**Pull request by Conorbro**
/config now reveals loaded Blackbox exporter configuration in YAML format.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -17,82 +17,82 @@ type Config struct {
...
 type SafeConfig struct {
-	sync.RWMutex
-	C *Config
+	sync.RWMutex `yaml:",omitempty"`
+	C            *Config `yaml:"config"`
 }
 
 type Module struct {
 	Prober  string        `yaml:"prober"`
 	Timeout time.Duration `yaml:"timeout"`
```
Comment (brian-brazil) [inline config.go]: this should be omit empty too

File: main.go [hunk 2]
```diff
@@ -176,15 +176,24 @@ func main() {
...
-            <h1>Blackbox Exporter</h1>
-            <p><a href="/probe?target=prometheus.io&module=http_2xx">Probe prometheus.io for http_2xx</a></p>
-            <p><a href="/metrics">Metrics</a></p>
-            </body>
-            </html>`))
+			<head><title>Blackbox Exporter</title></head>
+			<body>
+			<h1>Blackbox Exporter</h1>
+			<p><a href="/probe?target=prometheus.io&module=http_2xx">Probe prometheus.io for http_2xx</a></p>
+			<p><a href="/metrics">Metrics</a></p>
```
Comment (brian-brazil) [inline main.go]: Add a link to the config

File: main.go [hunk 3]
```diff
@@ -176,15 +176,24 @@ func main() {
...
+			</body>
+			</html>`))
+	})
+
+	http.HandleFunc("/config", func(w http.ResponseWriter, r *http.Request) {
+		c, err := yaml.Marshal(sc)
+		if err != nil {
+			log.Fatalf("Error marshalling configuration: %v", err)
```
Comment (brian-brazil) [inline main.go]: While this should be impossible, kill the process is probably a bit much. Log it and return a 500 I'd say.

**Comment by Conorbro** [comment]
@brian-brazil

File: blackbox.yml [hunk 4]
```diff
@@ -3,6 +3,10 @@ modules:
     prober: http
     timeout: 5s
     http:
+      method: POST
```
Comment (brian-brazil) [inline blackbox.yml]: This file is the default config, it shouldn't be changed for tests.

I'd put what you need in blackbox-good.yml

File: config.go [hunk 5]
```diff
@@ -17,82 +19,82 @@ type Config struct {
 }
 
 type SafeConfig struct {
-	sync.RWMutex
-	C *Config
+	sync.RWMutex `yaml:",omitempty"`
```
Comment (brian-brazil) [inline config.go]: Why are you adding yaml tags here?

File: config.go (hunk 1 above)
Comment (brian-brazil) [inline config.go]: This comment stands

File: config.go [hunk 6]
```diff
@@ -109,6 +111,14 @@ func checkOverflow(m map[string]interface{}, ctx string) error {
 	return nil
 }
 
+func (sc *SafeConfig) marshalYAML() ([]byte, error) {
```
Comment (brian-brazil) [inline config.go]: You don't need this, the default marshalling is fine

File: config_test.go [hunk 7]
```diff
@@ -25,3 +28,24 @@ func TestLoadBadConfig(t *testing.T) {
...
+		C: &Config{},
+	}
+
+	err := sc.reloadConfig("testdata/blackbox-good.yml")
+	if err != nil {
+		t.Errorf("Error loading config %v: %v", "testdata/blackbox-good.yml", err)
+	}
+
+	// String method must not reveal authentication credentials.
+	s, err := sc.marshalYAML()
```
Comment (brian-brazil) [inline config_test.go]: Marshall the config, not the wrapper around it.

**Comment by Conorbro** [comment]
@brian-brazil

File: main.go [hunk 8]
```diff
@@ -176,15 +176,27 @@ func main() {
...
+			<p><a href="/config">Configuration</a></p>
+			</body>
+			</html>`))
+	})
+
+	http.HandleFunc("/config", func(w http.ResponseWriter, r *http.Request) {
+		c, err := yaml.Marshal(sc.C)
+		if err != nil {
+			log.Warnf("Error marshalling configuration: %v", err)
+			http.Error(w, "", 500)
```
Comment (brian-brazil) [inline main.go]: Put the error in the http status also

File: main.go [hunk 9]
```diff
@@ -176,15 +176,27 @@ func main() {
...
+			<h1>Blackbox Exporter</h1>
+			<p><a href="/probe?target=prometheus.io&module=http_2xx">Probe prometheus.io for http_2xx</a></p>
+			<p><a href="/metrics">Metrics</a></p>
+			<p><a href="/config">Configuration</a></p>
+			</body>
+			</html>`))
+	})
+
+	http.HandleFunc("/config", func(w http.ResponseWriter, r *http.Request) {
+		c, err := yaml.Marshal(sc.C)
```
Comment (brian-brazil) [inline main.go]: You need to protect this read with the mutex

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**